trustedapp/src/main/java/org/dicekeys/trustedapp/view/DieFaceDrawables.kt

```diff
@@ -0,0 +1,176 @@
+package org.dicekeys.trustedapp.view
+
+import android.graphics.*
+import android.graphics.drawable.Drawable
+import androidx.core.content.res.ResourcesCompat
+import org.dicekeys.dicekey.Face
+import org.dicekeys.dicekey.FaceDimensionsFractional
+
+class Undoverline(val face: Face,
+                  val faceSize: Float,
+                  val isOverline: Boolean,
+                  val numberOfDots: Float = 11F,
+                  penColor: Int = Color.BLACK,
+                  holeColor: Int = Color.WHITE) : Drawable() {
+
+    val width: Float
+        get() = faceSize * FaceDimensionsFractional.undoverlineLength
+
+    val height: Float
+        get() = faceSize * FaceDimensionsFractional.undoverlineThickness
+
+    val code: Int
+        get() = (if (isOverline) face.overlineCode11Bits else face.underlineCode11Bits)?.toInt() ?: 0
+
+    val dotTop: Float
+        get() = faceSize * FaceDimensionsFractional.undoverlineMarginAlongLength
+
+    val marginAtStartAndEnd: Float
+        get() = faceSize * FaceDimensionsFractional.undoverlineMarginAtLineStartAndEnd
+
+    val dotStep: Float
+        get() = (width - 2 * marginAtStartAndEnd) / numberOfDots
+
+    val dotWidth: Float
+        get() = faceSize * (FaceDimensionsFractional.undoverlineDotWidth + 0.001F)
+
+    val dotHeight: Float
+        get() = faceSize * FaceDimensionsFractional.undoverlineDotHeight
+
+    val bitPositionsSet: List<Int>
+        get() {
+            val code = this.code
+            return (0..10).filter {
+                (code and (1 shl (10 - it))) != 0
+            }
+        }
+
+    private val penPaint = Paint().apply {
+        color = penColor
+    }
+
+    private val holePaint = Paint().apply {
+        color = holeColor
+    }
+
+    override fun draw(canvas: Canvas) {
+        val width = width
+        val height = height
+        val dotStep = dotStep
+        canvas.drawRect(0F, 0F, width, height, penPaint)
+        for (i in bitPositionsSet) {
+            val offsetX = marginAtStartAndEnd + i * dotStep
+            val offsetY = dotTop
+            canvas.drawRect(offsetX, offsetY, offsetX + dotWidth, offsetY + dotHeight, holePaint)
+        }
+    }
+
+    override fun setAlpha(alpha: Int) {
+        penPaint.alpha = alpha
+        holePaint.alpha = alpha
+    }
+
+    override fun setColorFilter(colorFilter: ColorFilter?) {
+        penPaint.colorFilter = colorFilter
+        holePaint.colorFilter = colorFilter
+    }
+
+    override fun getOpacity(): Int = PixelFormat.OPAQUE
+}
+
+class DieFaceUpright(val face: Face,
```

Comment:
This could be renamed DieFace.

The Swift code it's based on did not rotate the Face and so the face was always upright. Another component would render this component and then rotate it.

Since you are rotating the canvas here, this displays a correctly-oriented DieFace, and not an Upright dieface.  Hence,  it might make sense to merge this with DieFaceView